# Sequence alignment: what and why

We have decided that in this course, we will think of our three favorite biomolecules (DNA, RNA, protein) as strings/sequences, ignoring their secondary and 3D structures.
Naturally we want to be able to compare two (or more) sequences to identify similarities or dissimilarties between them.
For example, this need might arise when:

- Comparing human and chimpanzee genomes to figure out how and when we have diverged from our common ancestor
- Comparing normal human genome with a cancer genome to identify mutated, aberrant genes and their roles in the body's biochemistry
- Comparing DNA sequences collected from urban wastewater to antimicrobial resistant gene databases to discover the abundance and diversity of such resistant genes


## Can't we just use spell checkers?
Given large volumes of sequence data, we need a computer algorithm -- and a fast one -- for comparing sequences. 
To come up with an algorithm we need a computable definition for what similarity between two sequences means. 
Natural language processing, a field that bioinformatics borrows many ideas from, has developed to ways to compare sequences that occur in natural languages. Can't we just use say spell checkers on biological sequences?
To some extent, maybe, but not really, because the differences you might see between, say,  human and chimp genomes came about through a mechanism very different to how you make spelling errors when you are, say, typing messages on teeny tiny keyboard on a smartphone. 

## How do sequences evolve?
The main raw material for evolution is DNA **mutations**. 
These mutations spread (or not) through the population due to natural selection or just by chance (aka genetic drift). 


## Sequence alignment
The idea or the technique bioinformaticians have come up with to compare sequences is called sequence alignment.

### Global alignment

### Local alignment

## Important things to consider
### Algorithms
### Faster algorithms
### Scoring scheme

